subsys/bluetooth/host/conn.c

```diff
@@ -1725,6 +1725,35 @@ struct bt_conn *bt_conn_lookup_addr_le(u8_t id, const bt_addr_le_t *peer)
 	return NULL;
 }
 
+void bt_conn_foreach(int type, void (*func)(struct bt_conn *conn, void *data),
+		     void *data)
+{
+	int i;
+
+	for (i = 0; i < ARRAY_SIZE(conns); i++) {
+		if (!atomic_get(&conns[i].ref)) {
+			continue;
+		}
+
+		if (conns[i].type & type) {
+			continue;
+		}
+
+		func(bt_conn_ref(&conns[i]), data);
```

Comment:
I don't think it's particularly intuitive or consistent to force a reference upon the callback. It should be able to take a reference if it wants but I think we're risking reference leaks if it gets forced a reference that it's expected to unreformed before returning.

Reply:
Happy to loose the reference count, but I'm curious about why it is inconsistent. Mostly when a bt_conn pointer is returned it will be ref'ed. Which is why I assumed it would be necessary to do it here as well.

Reply:
None of our other callbacks (e.g. the connected/disconnected ones) force a new reference on the callback. When returning a pointer the situation is different, since then the function that returned it is no-longer on the stack, i.e. can't control the refcount itself.

This is mostly theoretical as long as we're just talking about cooperative threads, but when preemption comes into play we have to make sure that we don't get preempted and then some other thread invalidates the connection object. 

If the context you're doing this is at risk of being preempted, the code flow should be something like:
```
irq_lock();
if (!atomic_get(&conns[i].ref)) {
        irq_unlock();
        continue;
}

conn = bt_conn_ref(&conns[I]);
irq_unlock();

callback(conn);
bt_conn_unref(conn);
```

Since what you're adding is a new public API it may very well be called from a pre-emptible thread, so you might actually want something like the above for correctness. Since the stack itself will not drop the last reference for conn objects in ISR context (most of the stack runs in cooperative threads) k_sched_lock() might be preferable over irq_lock().

When dealing with functions that *return* a reference counted object the general design is that the function returns a new reference. That allows keeping the semantics same for all such functions while letting the function take care of safely acquiring that reference, i.e. either the caller gets NULL or then it gets a valid reference that it owns a reference count to.

Reply:
@jhedberg I agree. that looks like the correct way to handle it.
However I see that there are other places that should be fixed.
It seems that all `bt_conn_lookup_*` functions should have the same pattern.
They can all be called from application, either directly or indirectly through `bt_conn_create_*` functions.

Not sure what is best between irq_lock and k_sched_lock. It appears to be quite lightweight the code that should be in irq_lock, except when log level is causing print in `bt_conn_ref`.
If we use k_sched_lock it seems like it would be better to wrap it around the entire function, to avoid locking and unlocking inside the loop.

Should we fix this in this PR? Or create an issue and fix it separately?